### PR TITLE
DrivAerML: residual-prediction ablation on DM champion

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

`--residual-prediction` has never been tested in isolation for DrivAerML on the radford branch. The related PRs only test:
- casca #3192: `--asinh-pressure` alone
- himmel #3181: `--asinh-pressure --residual-prediction` combined

Neither isolates residual prediction as a single variable. This PR fills that gap with a clean ablation.

Residual prediction adds a learned correction on top of the freestream baseline, giving the model an easier regression target. On the noam branch this was part of the winning stack. If it helps DM independently of asinh-pressure, it should be mergeable on its own and composable with asinh downstream.

## Instructions

**Step 1 — 3-epoch smoke test (Trial 1 only):**

Run Trial 1 for 3 epochs first and report `val_primary/surface_rel_l2_pct` before committing to the full run.

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 3 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --residual-prediction \
  --wandb_group dm-residual-pred-ablation
```

If the 3-epoch run is not obviously diverging, proceed to the full trials below.

---

**Trial 1 — residual-prediction only (full run):**

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --residual-prediction \
  --wandb_group dm-residual-pred-ablation
```

**Trial 2 — residual-prediction + asinh-pressure (redundancy check vs himmel #3181):**

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --residual-prediction \
  --asinh-pressure --asinh-scale 1.0 \
  --wandb_group dm-residual-pred-ablation
```

Run Trial 2 in parallel with Trial 1 (different GPU).

**Target:** `val_primary/surface_rel_l2_pct` < **3.833%**

Report best val metric and W&B run ID for each trial. Note the epoch at which the best checkpoint was found.

## Baseline

Current DrivAerML best (PR #3072, student: eren):
- `val_primary/surface_rel_l2_pct`: **3.833%** (epoch 511)
- test `surface_rel_l2_pct`: 4.685%
- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, EMA=0.9995, gc=0.5, Fourier, no WD, no residual-prediction, no asinh

Reproduce command:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

External target for reference: AB-UPT ~3.71% (~500 epochs). Gap to close: 0.123 pp.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT franky:

## Results

### Trial 1 — `--residual-prediction` only

| Metric | Value |
|---|---|
| `val_primary/surface_rel_l2_pct` (best) | **15.001%** @ ep36 |
| Target | 3.833% |
| Gap to target | +11.17 pp |
| W&B run | `uteqp5p4` |
| Epochs completed | 70 (killed — catastrophic divergence at ep39) |
| Peak memory | not tracked |

**Command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --residual-prediction \
  --wandb-group dm-residual-pred-ablation --agent franky --wandb-name franky/dm-resid-pred-only
```

**What happened:** Residual prediction without asinh-pressure is catastrophically unstable on DrivAerML. The model appeared to converge initially (val dropping from ~14% at ep1 to ~9% at ep10), but gradient norms began spiking at ep39. By ep50, grad_norm was in the hundreds of thousands; by ep70 it reached millions. The model was fully diverged and unrecoverable. Killed at ep70 to free GPU.

---

### Trial 2 — `--residual-prediction --asinh-pressure --asinh-scale 1.0`

| Metric | Value |
|---|---|
| `val_primary/surface_rel_l2_pct` (best) | **3.999%** @ ep405 |
| Target | 3.833% |
| Gap to target | +0.17 pp |
| External ref (AB-UPT) | ~3.71% |
| Baseline (PR #3072, eren) | 3.833% val @ ep511, 4.685% test |
| W&B run | `6eud1yyg` |
| Epochs completed | 443 (killed — late-training divergence from ep416+) |
| Test metrics | Not available (process killed before test eval) |
| Peak memory | not tracked |

**Command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --residual-prediction --asinh-pressure --asinh-scale 1.0 \
  --wandb-group dm-residual-pred-ablation --agent franky --wandb-name franky/dm-resid-pred-asinh
```

**Convergence trajectory:**
- ep1-100: Rapid initial descent from ~14% to ~5.5%
- ep100-300: Steady improvement, ~5.5% → ~4.2%
- ep300-405: Slow convergence with plateau-then-breakthrough pattern at cosine cycle restarts, reaching 3.999%
- ep405-415: Stable but no further improvement
- ep416-443: Late-training destabilization — grad clip events escalating from single digits to 143/epoch, grad_norm rising from 0.07 to 0.53 (7x normal). Val metric degraded to ~4.5%

**What happened:** Residual prediction + asinh-pressure achieves **3.999%** — breaking below 4% but falling 0.17pp short of the 3.833% baseline target. The combination is stable for ~415 epochs but eventually destabilizes. The characteristic pattern of improvements at cosine cycle restarts (T_max=30) is very clear: each cycle restart yielded a new best val metric, suggesting the warm restarts are essential for escaping local minima with this configuration.

The result did NOT beat the baseline (3.833% @ ep511 from PR #3072). However, the baseline ran for 511 epochs without residual-prediction and without asinh-pressure — it's possible this config could improve further with more training time, but the late-training instability would need to be addressed first (lower LR, more aggressive grad clipping, or LR decay schedule modification).

---

### Bug fix in train.py

Fixed an `UnboundLocalError` on `primary_metric_key`: the local variable assignment `primary_metric_key = "val_primary/..."` shadowed the function `primary_metric_key()` at Python compile time, causing the function call on an earlier line to fail. Reordered the initialization block so the string is computed first, then assigned to `best_val_primary_metric_name`.

---

### Key findings

1. **Residual prediction alone is unstable on DrivAerML.** Without asinh-pressure, training diverges catastrophically around ep39. The freestream residual targets have extreme pressure values that cause gradient explosions.

2. **asinh-pressure is critical for stability.** It compresses the target distribution, preventing the extreme pressure values from dominating gradients. This is the enabling condition for residual prediction on DrivAerML.

3. **The combination (resid-pred + asinh) reaches 3.999% but doesn't beat the 3.833% baseline.** This suggests residual prediction does not independently improve DrivAerML performance on this branch — the baseline without it reaches a better val metric, just needing more epochs.

4. **Late-training instability is a concern.** Even with asinh-pressure, the training becomes unstable after ~415 epochs. This could be addressed with LR decay, warmup-adjusted grad clipping, or cosine schedule modifications.

### Suggested follow-ups

- **Try lower LR (2e-4 or 3e-4) with residual-prediction + asinh** to see if it stabilizes late training and converges further
- **Try longer T_max (60 or 100)** to give the model more time within each cosine cycle before restarts
- **Run the baseline for 511+ epochs with residual-prediction + asinh** to see if it can surpass 3.833% given sufficient time and stability
- **Investigate whether residual-prediction helps asinh-only** (compare casca #3192 asinh-only vs this Trial 2 asinh+resid-pred at matched epoch counts)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3193 because ADVISOR: Residual-prediction alone: 15.001% (crashed ep39, W&B: uteqp5p4) — catastrophic without asinh. Residual+asinh combined: 3.999% ep405 (W&B: 6eud1yyg) — tantalizingly close (+0.17pp) but destabilized ep416. Key finding: residual-prediction is a hard dependency on asinh for stability on DM, and even paired together doesn't beat the plain champion. The asinh transform tames the pressure gradients enough for residual to function, but the added complexity still introduces late-training instability. Casca #3192 (asinh-only) and himmel #3181 (asinh+residual) will provide the definitive comparison. Closing — residual-prediction on DM is not a standalone improvement..
